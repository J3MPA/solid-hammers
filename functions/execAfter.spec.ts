import execAfter from './execAfter'
describe('execAfter', () => {
  describe('Happy path', () => {
    it('should execute fn after nth time of execution', () => {
      const mockFn = jest.fn((s: string, n: number) => ({ s, n }))
      const fn = execAfter(3, mockFn)

      fn('first', 1)
      expect(mockFn).toHaveBeenCalledTimes(0)

      fn('second', 2)
      expect(mockFn).toHaveBeenCalledTimes(0)

      fn('third', 3)
      expect(mockFn).toHaveBeenCalledTimes(0)

      fn('forth', 4)
      expect(mockFn).toHaveBeenCalledTimes(1)
      expect(mockFn).toHaveBeenCalledWith('forth', 4)
      expect(mockFn).toHaveReturnedTimes(1)
      expect(mockFn).toHaveReturnedWith({ s: 'forth', n: 4 })
    })
  })

  describe('Sad path', () => {
    it('should throw TypeError if nth time is not a real number', () => {
      const mockFn = jest.fn()
      expect(() => execAfter(NaN, mockFn)).toThrowError(TypeError)
      expect(() => execAfter(NaN, mockFn)).toThrowErrorMatchingSnapshot()
      expect(() => execAfter(Infinity, mockFn)).toThrowError(TypeError)
      expect(() => execAfter(Infinity, mockFn)).toThrowErrorMatchingSnapshot()
      expect(() => execAfter(null as any, mockFn)).toThrowError(TypeError)
      expect(() =>
        execAfter(null as any, mockFn)
      ).toThrowErrorMatchingSnapshot()
      expect(() => execAfter(undefined as any, mockFn)).toThrowError(TypeError)
      expect(() =>
        execAfter(undefined as any, mockFn)
      ).toThrowErrorMatchingSnapshot()
    })
  })
})
