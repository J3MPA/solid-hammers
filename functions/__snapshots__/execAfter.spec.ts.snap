// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execAfter Sad path should throw TypeError if nth time is not a real number 1`] = `"nthTime must be a real number"`;

exports[`execAfter Sad path should throw TypeError if nth time is not a real number 2`] = `"nthTime must be a real number"`;

exports[`execAfter Sad path should throw TypeError if nth time is not a real number 3`] = `"nthTime must be a real number"`;

exports[`execAfter Sad path should throw TypeError if nth time is not a real number 4`] = `"nthTime must be a real number"`;
